packages/valory/skills/abstract_round_abci/models.py

```diff
@@ -160,18 +364,58 @@ def _ensure_setup(self, necessary_keys: Iterable[str]) -> Any:
         enforce(found, fail_msg)
 
 
-class SharedState(Model):
+class _MetaSharedState(ABCMeta):
+    """A metaclass that validates SharedState's attributes."""
+
+    def __new__(mcs, name: str, bases: Tuple, namespace: Dict, **kwargs: Any) -> Type:  # type: ignore
+        """Initialize the class."""
+        new_cls = super().__new__(mcs, name, bases, namespace, **kwargs)
+
+        if ABC in bases:
+            # abstract class, return
+            return new_cls
+        if not issubclass(new_cls, SharedState):
+            # the check only applies to SharedState subclasses
+            return new_cls
+
+        mcs._check_consistency(cast(Type[SharedState], new_cls))
+        return new_cls
+
+    @classmethod
+    def _check_consistency(mcs, shared_state_cls: Type["SharedState"]) -> None:
+        """Check consistency of class attributes."""
+        mcs._check_required_class_attributes(shared_state_cls)
+
+    @classmethod
+    def _check_required_class_attributes(
+        mcs, shared_state_cls: Type["SharedState"]
+    ) -> None:
+        """Check that required class attributes are set."""
+        if not hasattr(shared_state_cls, "abci_app_cls"):
+            raise AttributeError(f"'abci_app_cls' not set on {shared_state_cls}")
+        abci_app_cls = shared_state_cls.abci_app_cls
+        if not inspect.isclass(abci_app_cls):
+            raise AttributeError(f"The object `{abci_app_cls}` is not a class")
+        if not issubclass(abci_app_cls, AbciApp):
+            cls_name = AbciApp.__name__
+            cls_module = AbciApp.__module__
+            raise AttributeError(
+                f"The class {abci_app_cls} is not an instance of {cls_module}.{cls_name}"
+            )
+
+
+class SharedState(Model, ABC, metaclass=_MetaSharedState):  # type: ignore
```

Comment:
Similar to some of our other classes, we move the checks into the metaclass to catch issues earlier. Also means we save boilerplate on the concrete skills.